brainiak/searchlight/searchlight.py

```diff
@@ -0,0 +1,344 @@
+        self.fn = fn
+
+    def _get_subarray(self, data, idx):
+        """ Return a subarray with radius 'rad'
+
+        The subarray is centered around 'idx' with the
+        same list structure as 'data'.
+
+        Parameters
+        ----------
+        data: A list of 4D numpy.ndarray objects which must have at least one
+        element. The computation is applied to this data.
+
+        idx: Three-element tuple containing the current center of the
+        searchlight
+
+        Returns
+        -------
+        A list of 4D numpy ndarrays (subarrays)
+
+        """
+
+        def _list_slice(l):
+            if(isinstance(l, list)):
+                return [_list_slice(el) for el in l]
+            else:
+                return l[:, idx[0] - self.rad:idx[0] + self.rad + 1,
+                         idx[1] - self.rad:idx[1] + self.rad + 1,
+                         idx[2] - self.rad:idx[2] + self.rad + 1]
+
+        return _list_slice(data)
+
+    def _get_submask(self, mask, idx):
+        """ Return a subarray of the mask centered around 'idx'
+
+        Parameters
+        ----------
+        mask: A 3D numpy.ndarray object with type np.bool indicating which
+                    voxels are active in the searchlight operation.
+
+        idx: Three-element tuple containing the current center of the
+        searchlight
+
+        Returns
+        ------
+        A 3D numpy ndarray (subarray)
+
+        """
+
+        return mask[idx[0] - self.rad:idx[0] + self.rad + 1,
+                    idx[1] - self.rad:idx[1] + self.rad + 1,
+                    idx[2] - self.rad:idx[2] + self.rad + 1]
+
+    def _do_master(self, tasks, data, mask, bcast_var):
+        """ Distribute tasks dynamically to ranks 1-size (master)
+
+        Parameters
+        ----------
+        tasks: searchlight centers, a list of tuples
+
+        data: A list of 4D numpy.ndarray objects which must have at least one
+        element. The computation is applied to this data.
+
+        mask: A 3D numpy.ndarray object with type np.bool indicating which
+        voxels are active in the searchlight operation.
+
+        bcast_var: An object of any pickle-able type which is included
+        as a parameter to each invocation of the user-specified function
+
+        Returns
+        -------
+        A 3D numpy ndarray containing the outputs of the function at
+        each voxel.
+
+        """
+
+        comm = MPI.COMM_WORLD
+        size = comm.Get_size()
+        results = []
+
+        # If there are no workers, then do everything here
+        if size == 1:
+            for idx in tasks:
+                results += [(idx, self.fn(self._get_subarray(data, idx),
+                                          self._get_submask(mask, idx),
+                                          bcast_var))]
+        # Assign tasks to workers
+        else:
+            num_active = 0
+
+            # Send wait for requests from workers
+            for idx in tasks:
```

Comment:
The log goes to either stdout or a file, which you can set by `logging.basicConfig` in your running code, e.g. in the example code.